File: Populating Next Right Pointers in Each Node.cpp
class Solution {
public:
    void connect(TreeLinkNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        TreeLinkNode *father = root;
       
        while(father)
        {
            TreeLinkNode *current = father;
            TreeLinkNode *prev = NULL;
            TreeLinkNode *child = NULL;
            while(current)
            {
                if(current->left)
                {
                    if(child==NULL)
                        child = current->left;
                        
                    if(prev==NULL)
                    {
                        prev = current->left;
                    }
                    else
                    {
                        prev->next = current->left;
                        prev = current->left;
                    }
                }
                
                if(current->right)
                {
                    if(child==NULL)
                        child = current->right;
                    
                    if(prev==NULL)
                    {
                        prev = current->right;
                    }
                    else
                    {
                        prev->next = current->right;
                        prev = current->right;
                    }
                }
                
                current = current->next;
            }
            
            father = child;
        }
    }
};

File: Populating Next Right Pointers in Each Node.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    
    public TreeLinkNode preNode;
    public TreeLinkNode nextLevel;
    
    public void connect(TreeLinkNode root) {
        // Start typing your Java solution below
        // DO NOT write main() function
        preNode = null;
        nextLevel = null;
        TreeLinkNode curNode = root;
        while(curNode != null)
        {
            connectNode(curNode.left);
            connectNode(curNode.right);
            if(curNode.next == null)
            {
                curNode = nextLevel; 
                preNode = null;
                nextLevel = null;
            }
            else
            {
                curNode = curNode.next;
            }
        }
    }
    
    private void connectNode(TreeLinkNode node)
    {
        if(node!= null)
        {
            if(nextLevel == null) 
            {
                nextLevel = node;
            }
            if(preNode != null)
            {
                preNode.next = node;
            }
            preNode = node;
        }
    }
}